### PR TITLE
schema changes with election related endpoint 

### PR DESCRIPTION
**Pull request by MrClemRkz**

### CODE REVIEW

File: src/routes/electionRouter.js [hunk 1]
```diff
@@ -1,28 +1,29 @@
...
-module.exports = router;
\ No newline at end of file
+export const initElectionRouter = (app) => {
+	electionRouter(app, [
+		{
+			method: GET,
+			path: '/elections/:electionId',
+			schema: GET_ELECTION_BY_ID_SCHEME,
+			handler: (req, res, next) => {
+				// return ElectionService.getElectionById(req)
```
Comment (anandafit) [inline src/routes/electionRouter.js]: remove the commented code

File: src/routes/nominationRouter.js [hunk 2]
```diff
@@ -76,6 +73,18 @@ router.get('/:nomination_id/candidate', (req, res, next) => {
...
 });
 
+/** 
+ * 8th to put  added or updated nominated candidate details */
+router.put('/:nomination_id/candidate/add', (req, res, next) => {
+    return CandidateService.updateNominationCandidates(req).then((results) => {
+        if (result instanceof Error) {
+            next(result);
+        } else {
+            res.send(result);
```
Comment (anandafit) [inline src/routes/nominationRouter.js]: status code missing

File: src/routes/nominationRouter.js [hunk 3]
```diff
@@ -104,32 +113,5 @@ router.put('/:nomination_id/payment/update', (req, res, next) => {
...
-
-*/
-router.put('/add_candidates', (req, res) => {
-    return NominationService.updateNominationCandidates(req).then((results) => {
-        res.json(results);
-    });
-});
-
 
 module.exports = router;
```
Comment (anandafit) [inline src/routes/nominationRouter.js]: keep new line end of file

File: src/routes/schema/electionSchema.js [hunk 4]
```diff
@@ -0,0 +1,5 @@
+import Joi from 'joi';
+
+export const GET_ELECTION_BY_ID_SCHEME = Joi.object().keys({
+    max: Joi.string().max(36),
+});
```
Comment (anandafit) [inline src/routes/schema/electionSchema.js]: keep new line end of file

File: src/routes/schema/electionSchema.js [hunk 5]
```diff
@@ -0,0 +1,5 @@
+import Joi from 'joi';
+
+export const GET_ELECTION_BY_ID_SCHEME = Joi.object().keys({
+    max: Joi.string().max(36),
```
Comment (anandafit) [inline src/routes/schema/electionSchema.js]: do you really want to max param here

File: src/routes/schema/userSchema.js [hunk 6]
```diff
@@ -2,9 +2,9 @@ import Joi from 'joi';
 
 export const USER_EDIT_SCHEMA = Joi.object().keys({
   name: Joi.string().max(255).required(true),
-  id: Joi.number(),
+  id: Joi.string().max(255).required(true),
```
Comment (anandafit) [inline src/routes/schema/userSchema.js]: 1. You would validate for max, min, allow characters.
2. name and id are the only available pos params

File: src/service/electionService.js [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        }
+    } catch (error) {
+        throw error;
+    }
+}
+
+export default {
+    getElectionById,
+    getElectionWithTimelineById,
+}
```
Comment (anandafit) [inline src/service/electionService.js]: new line end of file

File: src/service/electionService.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    try {
+        const id = req.params.electionId;
+        const election = await ElectionRepo.getbyId(id);
+        if (!_.isEmpty(election)){
+            return ElectionManager.mapToElectionModel(election);
+        } else {
+            throw new ApiError("Election not found");
+        }
+    } catch (error) {
+        throw error;
```
Comment (anandafit) [inline src/service/electionService.js]: Throw the specific error with error code

File: src/service/electionService.js [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    try {
+        const id = req.params.electionId;
+        const election = await ElectionRepo.getbyIdWithTimelineData(id);
+        if(!_.isEmpty(election)){
+            return ElectionManager.mapToElectionModelWithTimeline(election);
+        } else {
+            throw new ApiError("Election not found");
+        }
+    } catch (error) {
+        throw error;
```
Comment (anandafit) [inline src/service/electionService.js]: Throw a specific error with error code

File: src/service/index.js [hunk 10]
```diff
@@ -3,11 +3,13 @@ import NominationService from './nominationService';
...
+import ElectionService from './electionService';
 
 export {
   UserService,
   PaymentService,
   NominationService,
   CandidateService,
   TeamService,
+  ElectionService,
 }
```
Comment (anandafit) [inline src/service/index.js]: new line please

File: src/service/paymentService.js [hunk 11]
```diff
@@ -1,40 +1,42 @@
 import _ from 'lodash';
```
Comment (anandafit) [inline src/service/paymentService.js]: Noticed that you don't have handle error with try catch. Please do so and throw error massage and error code accordingly

File: src/manager/election/electionManager.js [hunk 12]
```diff
@@ -0,0 +1,48 @@
+import { Election } from 'Models';
```
Comment (anandafit) [inline src/manager/election/electionManager.js]: Please unit test for each manager functions

File: src/manager/index.js [hunk 13]
```diff
@@ -2,10 +2,12 @@ import UserManager from './user/userManager';
 import TeamManager from './team/teamManager';
 import NominationEntryManager from './nomination/nominationEntryManager';
 import {PaymentManger} from './payment/paymentManager';
+import ElectionManager from './election/electionManager';
 
 export {
   UserManager,
```
Comment (anandafit) [inline src/manager/index.js]: Please add unit test for each manager functions

File: src/model/User.js [hunk 14]
```diff
@@ -2,6 +2,6 @@
 import {Record} from 'typed-immutable';
 
 export const User =  Record({
-  id: Number(),
+  id: String(),
   name: String(),
 });
```
Comment (anandafit) [inline src/model/User.js]: new line

File: src/repository/election.js [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+			throw new DBError(error);
+		});
+};
+
+
+
+export default {
+	getbyId,
+	getbyIdWithTimelineData,
+}
```
Comment (anandafit) [inline src/repository/election.js]: new line

File: src/routes/nominationRouter.js [hunk 16]
```diff
@@ -1,5 +1,5 @@
 import _ from 'lodash';
```
Comment (anandafit) [inline src/routes/nominationRouter.js]: This router is still in the old format. if you don't want this router, remove that.

File: src/manager/election/electionManager.js [hunk 17]
```diff
@@ -0,0 +1,48 @@
...
+
+
+const mapToElectionModel = (election) => {
+	const mappedElection = joinjs.map(election, resultMaps, 'electionMap', 'election_');
+
+	//just an example how to convert row data object to Model
+	return Election({
+		id: mappedElection[0].id,
+		name: mappedElection[0].name,
+		module_id: mappedElection[0].module_id,
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/manager/election/electionManager.js]: done

File: src/manager/election/electionManager.js [hunk 18]
```diff
@@ -0,0 +1,48 @@
...
+	});
+};
+
+const mapToElectionModelWithTimeline = (election) => {
+	const mappedElection = joinjs.map(election, resultMaps, 'electionMap', 'election_');
+
+	//just an example how to convert row data object to Model
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/manager/election/electionManager.js]: done

File: src/manager/election/electionManager.js [hunk 19]
```diff
@@ -0,0 +1,48 @@
...
+	return Election({
+		id: mappedElection[0].id,
+		name: mappedElection[0].name,
+		module_id: mappedElection[0].module_id,
+	});
+};
+
+const mapToElectionModelWithTimeline = (election) => {
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/manager/election/electionManager.js]: done

File: src/manager/election/electionManager.js [hunk 20]
```diff
@@ -0,0 +1,48 @@
...
+	// return Election({
+	// 	id: mappedElection[0].id,
+	// 	name: mappedElection[0].name,
+	// 	module_id: mappedElection[0].module_id,
+	// });
+	return mappedElection[0];
+}
+
+export default {
+	mapToElectionModel,
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/manager/election/electionManager.js]: nope. one was left for it's been used in electionService method and all from electionRouter. I have commented them all.

File: src/repository/election.js [hunk 21]
```diff
@@ -0,0 +1,51 @@
...
+import { formatQueryToBulkInsert, formatDataToBulkInsert } from './sqlHelper';
+
+const ELECTION_SELECT_QUERY = `SELECT 
+ID AS election_id,
+NAME AS election_name,
+MODULE_ID AS election_module_id
+FROM ELECTION
+WHERE ID = :id`;
+
+const getbyId = (electionId) => {
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/repository/election.js]: This I did for it to have a different method name from electionService.js file. Anyhow as we discussed will have 'fetch' on Repository js files and 'get' on Service js files.

File: src/repository/election.js [hunk 22]
```diff
@@ -0,0 +1,51 @@
...
+const ELECTION_WITH_TIMELINE_SELECT_QUERY = `SELECT 
+e.ID AS election_id,
+e.NAME AS election_name,
+e.MODULE_ID AS election_module,
+etc.KEY_NAME AS timeline_key,
+etcd.value AS timeline_value
+FROM ELECTION e
+	INNER JOIN ELECTION_TIMELINE_CONFIG_DATA etcd ON e.ID = etcd.ELECTION_ID
+	INNER JOIN ELECTION_TIMELINE_CONFIG etc ON etc.ID = etcd.ELECTION_TIMELINE_CONFIG_ID
+WHERE e.ID = :id`
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/repository/election.js]: done

File: src/model/Election.js [hunk 23]
```diff
@@ -0,0 +1,8 @@
+
+import {Record} from 'typed-immutable';
+
+export const Election =  Record({
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/model/Election.js]: done.

[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/model/Election.js]: changed the import to:
import {Record, List} from 'typed-immutable';

File: src/repository/election.js [hunk 24]
```diff
@@ -0,0 +1,51 @@
...
+
+
+const ELECTION_WITH_TIMELINE_SELECT_QUERY = `SELECT 
+e.ID AS election_id,
+e.NAME AS election_name,
+e.MODULE_ID AS election_module,
+etc.KEY_NAME AS timeline_key,
+etcd.value AS timeline_value
+FROM ELECTION e
+	INNER JOIN ELECTION_TIMELINE_CONFIG_DATA etcd ON e.ID = etcd.ELECTION_ID
```
[the remark this answers is not in the document]
Comment (MrClemRkz) [reply src/repository/election.js]: changed to LEFT JOIN

File: src/routes/schema/electionSchema.js (hunk 4 above)
Comment (MrClemRkz) [reply src/routes/schema/electionSchema.js]: done

File: src/routes/nominationRouter.js (hunk 3 above)
Comment (MrClemRkz) [reply src/routes/nominationRouter.js]: nomination files will be amended later.

File: src/routes/schema/electionSchema.js (hunk 5 above)
Comment (MrClemRkz) [reply src/routes/schema/electionSchema.js]: not really. just has adhered to the coding pattern.

File: src/routes/schema/userSchema.js (hunk 6 above)
Comment (MrClemRkz) [reply src/routes/schema/userSchema.js]: noted.

File: src/service/electionService.js (hunk 7 above)
Comment (MrClemRkz) [reply src/service/electionService.js]: done

File: src/service/electionService.js (hunk 8 above)
Comment (MrClemRkz) [reply src/service/electionService.js]: threw ServerError

File: src/service/electionService.js (hunk 9 above)
Comment (MrClemRkz) [reply src/service/electionService.js]: done

File: src/service/index.js (hunk 10 above)
Comment (MrClemRkz) [reply src/service/index.js]: done

File: src/service/paymentService.js (hunk 11 above)
Comment (MrClemRkz) [reply src/service/paymentService.js]: will be update on yujith's fork.

File: src/model/User.js (hunk 14 above)
Comment (MrClemRkz) [reply src/model/User.js]: done

File: src/repository/election.js (hunk 15 above)
Comment (MrClemRkz) [reply src/repository/election.js]: done

File: src/routes/nominationRouter.js (hunk 16 above)
Comment (MrClemRkz) [reply src/routes/nominationRouter.js]: will be updated later.

File: src/manager/index.js (hunk 13 above)
Comment (anandafit) [reply src/manager/index.js]: added todo to add this latter

**Pull request merged**